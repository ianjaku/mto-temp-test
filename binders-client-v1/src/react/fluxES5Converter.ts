/* eslint-disable @typescript-eslint/no-explicit-any */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function fixES5FluxContainer<T extends new (...args: any[]) => any>(
    containerClass: T
): T {
    const tmp = containerClass;
    const newClass = function(...args: any[]) {
        return new tmp(...args);
    } as unknown as T;
    newClass.prototype = tmp.prototype;
    (newClass as any).getStores = (tmp as any).getStores;
    (newClass as any).calculateState = (tmp as any).calculateState;
    return newClass;
}
