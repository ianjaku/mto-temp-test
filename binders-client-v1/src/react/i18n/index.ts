import i18n from "../../i18n";
export { useTranslation, withTranslation } from "react-i18next";
export { type TFunction } from "../../i18n";
export default i18n;
