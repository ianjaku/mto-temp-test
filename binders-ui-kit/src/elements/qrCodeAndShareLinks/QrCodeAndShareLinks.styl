@require "../../vars.styl"

.qrCodeAndShareLinks
	display: block
	width: 100%
	padding: 16px
	box-sizing: border-box

	.button
		width: 100%
		margin: 0 0 5px
		border-width: 1px
		box-sizing: border-box

	.dropdown
		width: 100% !important
		margin: 5px 0 15px
		box-sizing: border-box

	&-linkWrapper
		position: relative
		display: block
		margin: 15px 0
		padding: 5px 10px
		box-sizing: border-box
		border: 1px solid borderGrayColor

	&-link
		display: inline-block
		position: relative
		max-width: calc(100% - 30px)
		box-sizing: border-box
		vertical-align: middle
		color: darkGrayColor
		font-size: 13px
		line-height 20px
		overflow: hidden
		white-space: nowrap
		text-overflow: ellipsis

	&-clipboard
		display: inline-block

	&-icon
		display: inline-block
		position: relative
		bottom: 0
		right: -10px
		cursor: pointer
		vertical-align: middle

		&:hover
			opacity: .75

