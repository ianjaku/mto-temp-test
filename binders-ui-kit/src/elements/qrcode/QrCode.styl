.qrCode
    &-dataImage
        display: block
        width: 100%
        margin: 5px auto 0
    &-dataLogo
        display: none