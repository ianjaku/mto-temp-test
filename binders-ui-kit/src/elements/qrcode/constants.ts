export const QR_CODE_VERSION = 4;
export const LOGO_RATIO = 4.3; // MT-3317
