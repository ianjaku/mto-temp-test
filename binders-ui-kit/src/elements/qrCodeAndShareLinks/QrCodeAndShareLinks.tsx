import Dropdown, { IDropdownElement } from "../dropdown";
import Button from "../button";
import ClipboardIcon from "../icons/CopyClipboard";
import { CopyToClipboard } from "react-copy-to-clipboard";
import { QrCode } from "../qrcode/QrCode";
import React from "react";
import { TranslationKeys as TK } from "@binders/client/lib/react/i18n/translations";
import { useTranslation } from "@binders/client/lib/react/i18n";
import "./QrCodeAndShareLinks.styl";

export interface ShareLanguage {
    id: number;
    languageLabel: string;
    link: string;
}

interface Props {
    linkOptions?: string[];
    onLinkSelected?: (link: string) => void;
    link: string;
    languages?: ShareLanguage[];
    selectedItemId?: number;
    languageSelect?: (id: number) => void;
    showFlashMessage?: () => void;
    useLogo?: boolean;
    canvasMaxWidth?: number;
}

export const QrCodeAndShareLinks: React.FC<Props> = (props) => {

    const [downloadableImage, setDownloadableImage] = React.useState<string>();
    const { t } = useTranslation();

    const onCopyLink = (_txt, success) => {
        if (props.showFlashMessage && success) {
            props.showFlashMessage();
        }
    }

    const dropdownElements = (props.languages || []).map(language => ({
        id: language.id,
        label: language.languageLabel,
    }));


    return (
        <div className="qrCodeAndShareLinks">
            {dropdownElements.length > 1 &&
                <Dropdown
                    type="Languages"
                    elements={dropdownElements}
                    maxRows={5}
                    selectedElementId={props.selectedItemId}
                    onSelectElement={props.languageSelect}
                />}

            {props.linkOptions && (
                <Dropdown
                    type="Languages"
                    elements={props.linkOptions.map<IDropdownElement>((link) => ({
                        id: link,
                        label: link,
                    }))}
                    maxRows={5}
                    selectedElementId={props.link}
                    onSelectElement={props.onLinkSelected}
                    hideSelectedElementInList
                />
            )}
            <QrCode
                link={props.link}
                useLogo={props.useLogo}
                canvasMaxWidth={props.canvasMaxWidth}
                setDownloadableImage={setDownloadableImage}
            />
            <Button
                hrefAnchor={downloadableImage}
                downloadableName="qrcode.png"
                text={`${t(TK.General_Download)} .png`}
                secondary={true}
            />
            <div className="qrCodeAndShareLinks-linkWrapper">
                <span className="qrCodeAndShareLinks-link">{props.link}</span>
                <CopyToClipboard text={props.link} onCopy={onCopyLink}>
                    <div className="qrCodeAndShareLinks-icon">
                        <ClipboardIcon testId="qrCodeAndShareLinks-clipboardIcon" />
                    </div>
                </CopyToClipboard>
            </div>
        </div>
    );
}
