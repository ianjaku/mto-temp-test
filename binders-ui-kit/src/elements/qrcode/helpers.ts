import * as QRCode from "qrcode";
import { LOGO_RATIO, QR_CODE_VERSION } from "./constants";
import { QRCodeRenderersOptions } from "qrcode";

export const drawQrOnCanvas = (
    link: string,
    imgElement: HTMLImageElement,
    canvasElement?: HTMLCanvasElement,
    canvasMaxWidth?: number,
    useLogo = false,
    versionOverride?: number,
    secondAttempt = false,
): Promise<HTMLCanvasElement> => {

    const options: QRCodeRenderersOptions = {
        errorCorrectionLevel: useLogo ? "M" : "L",
        margin: 1,
        version: versionOverride || QR_CODE_VERSION,
        width: calculateCanvasWidth(canvasElement, canvasMaxWidth),
    };

    const canvas = canvasElement || document.createElement("canvas") as HTMLCanvasElement;
    return new Promise<HTMLCanvasElement>((resolve, reject) => {
        QRCode.toCanvas(canvas, link, options, (err) => {
            if (!err) {
                if (useLogo) {
                    drawScaledLogo(imgElement, canvas, Math.round(canvas.width / LOGO_RATIO));
                }
                return resolve(canvas);
            }
            if (!secondAttempt && err.message.indexOf("Minimum version required") > -1) {
                const lines = err.message.split("\n");
                const versionLine = lines.find(l => l.indexOf("Minimum version required") === 0);
                const requiredVersion = versionLine.substring(versionLine.indexOf(": ") + 1);
                const version = parseInt(requiredVersion, 10);
                return resolve(drawQrOnCanvas(
                    link,
                    imgElement,
                    canvasElement,
                    canvasMaxWidth,
                    useLogo,
                    version,
                ));
            }
            reject(err);
        });
    });
};

const isImageLoaded = (img: HTMLImageElement): boolean => {
    return img.complete && (typeof img.naturalWidth === "undefined" || img.naturalWidth !== 0);
};

const scaleImage = (
    source: HTMLImageElement,
    width: number,
    height: number
): HTMLCanvasElement => {
    const tempCanvas = document.createElement("canvas");
    tempCanvas.width = width;
    tempCanvas.height = height;
    drawImage(source, tempCanvas.getContext("2d"), 0, 0, width, height);
    return tempCanvas;
};

const drawScaledLogo = (
    imgElement: HTMLImageElement,
    canvas: HTMLCanvasElement,
    logoSize: number
): void => {
    const scaledLogo = scaleImage(imgElement, logoSize * 2, logoSize * 2);
    const ctx = canvas.getContext("2d");
    if (ctx) {
        ctx.drawImage(
            scaledLogo,
            (canvas.width / 2) - (logoSize / 2),
            (canvas.width / 2) - (logoSize / 2),
            logoSize,
            logoSize,
        );
    }
};

const drawImage = (
    source: HTMLImageElement,
    ctx: CanvasRenderingContext2D | null,
    x: number,
    y: number,
    width: number,
    height: number,
): void => {
    if (ctx && isImageLoaded(source)) {
        ctx.drawImage(source, x, y, width, height);
        return;
    }
    source.onload = () => {
        if (ctx) {
            ctx.drawImage(source, x, y, width, height);
        }
    }
};

const calculateCanvasWidth = (canvasParentElement: HTMLElement, canvasMaxWidth?: number): number => {
    const { paddingLeft, paddingRight } = getComputedStyle(canvasParentElement);
    const padding = parseFloat(paddingLeft) + parseFloat(paddingRight);
    const width = canvasParentElement.clientWidth - padding;
    return Math.min(width, canvasMaxWidth ?? 2000);
};
