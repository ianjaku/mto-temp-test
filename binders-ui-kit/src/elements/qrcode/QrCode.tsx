import React, { useCallback, useEffect, useRef } from "react";
import QR_CODE_LOGO from "../../../public/m-logo.svg";
import { drawQrOnCanvas } from "./helpers";
import { usePrevious } from "@binders/client/lib/react/helpers/hooks";
import "./QrCode.styl";

interface Props {
    link: string;
    useLogo?: boolean;
    canvasMaxWidth?: number;
    setDownloadableImage?: (downloadableImage: string) => void;
    setCopyableBlob?: (blob: Blob) => void;
    style?: React.CSSProperties;
}

export const QrCode: React.FC<Props> = ({ link, useLogo, canvasMaxWidth, setDownloadableImage, setCopyableBlob, style }) => {
    const qrCodeCanvasRef = useRef<HTMLCanvasElement>(null);
    const imgRef = useRef<HTMLImageElement>(null);
    const prevLink = usePrevious(link);

    const onLoadImgAsync = useCallback(async () => {
        try {
            const canvas = await drawQrOnCanvas(
                link,
                imgRef.current,
                qrCodeCanvasRef.current,
                canvasMaxWidth,
                useLogo
            );
            if (setDownloadableImage) {
                setDownloadableImage(canvas.toDataURL("image/png", 1))
            }
            if (setCopyableBlob) {
                canvas.toBlob(blob => {
                    if (blob) {
                        setCopyableBlob(blob);
                    }
                });
            }
        } catch (error) {
            // eslint-disable-next-line no-console
            console.error(error);
        }
    }, [canvasMaxWidth, link, setDownloadableImage, setCopyableBlob, useLogo])

    useEffect(() => {
        if (link && prevLink && link !== prevLink) {
            onLoadImgAsync();
        }
    }, [link, prevLink, onLoadImgAsync])

    return (
        <div className="qrCode" style={style || {}}>
            <img
                style={{ display: "none" }}
                className="qrCode-dataLogo"
                ref={imgRef}
                src={QR_CODE_LOGO}
                onLoad={onLoadImgAsync}
            />
            <canvas
                className="qrCode-dataImage"
                ref={qrCodeCanvasRef}
            />
        </div>
    )
}
