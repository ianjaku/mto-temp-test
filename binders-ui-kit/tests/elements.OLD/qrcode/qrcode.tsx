import * as QRCodeAPI from "qrcode";
import * as React from "react";
import { QrCodeAndShareLinks } from "../../../src/elements/qrCodeAndShareLinks/QrCodeAndShareLinks";
import { create } from "react-test-renderer";
import { shallow } from "enzyme";

const link = "http://manual.to/";

const createQRCode = () => (
    <QrCodeAndShareLinks link={link} />
);

function asyncGenerateQRCode(url: string): Promise<unknown> {

    const options = {
        margin: 0,
    };

    return new Promise((resolve, reject) => {
        // eslint-disable-next-line @typescript-eslint/ban-types
        QRCodeAPI.toDataURL(url, options, (error: object, dataUrl: string) => {
            if (error) {
                reject(error);
            } else {
                const QRObject = {
                    dataImage: dataUrl,
                    link: url,
                };
                resolve(QRObject);
            }
        });
    });
}


describe("QRCode", () => {

    test("Base tabs (snapshot)", () => {
        return asyncGenerateQRCode(link).then(() => {
            const QRComponent = create(createQRCode());
            const serialized = QRComponent.toJSON();
            expect(serialized).toMatchSnapshot();
        });
    });

    test("Check QR Code Async loaded", () => {
        const QRCodeMount = shallow(createQRCode(), { lifecycleExperimental: true });
        return asyncGenerateQRCode(link).then((data) => {
            expect(QRCodeMount.state("object")).toEqual(data);
        });
    });
});
