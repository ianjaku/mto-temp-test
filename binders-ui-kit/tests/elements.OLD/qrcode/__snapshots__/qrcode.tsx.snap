// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCode Base tabs (snapshot) 1`] = `
<div
  className="qrcode-content"
>
  <img
    alt="QR Code download"
    className="qrcode-dataImage"
    src={undefined}
  />
  <div
    className="qrcode-linkWrapper"
  >
    <span
      className="qrcode-link"
    >
      http://manual.to/
    </span>
    <div
      className="qrcode-icon"
      onClick={[Function]}
    >
      <img
        src=""
      />
    </div>
  </div>
  <div
    className="button button--secondary"
    onBlur={undefined}
    onClick={[Function]}
    tabIndex={undefined}
  >
    Download .png
    
    <a
      className="button-link"
      download="qrcode.png"
      href="image/png"
    />
  </div>
</div>
`;
